SSH and FTP deployment + dot notation version increments

tanis2000 opened the pull request:
Hello,

I've added SSH and FTP deployment targets. 
I've also added support for smart version number increments that takes care of version numbers with dot notation.
I hope you can merge them in the main project.
Thanks!

tanis2000 committed:
Added basic support for web deployment through FTP and SSH

tanis2000 committed:
Changed SSH to use Net::SSH

tanis2000 committed:
Added documentation for web deployment

tanis2000 committed:
Added support to incrementing build number in plist taking care of standard int values and dot notation version numbers.

tanis2000 committed:
Fixed a small bug when calculating the next build number

rayh commented on lib/xcode/deploy/web.rb:
```diff
@@ -0,0 +1,141 @@
+
+module Xcode
+  module Deploy
+    class Web
```
Not sure if Web is the right name...  Could we perhaps have a class per protocol and push the plist/html generation into some utility classes (e.g., It would be nice to have S3 as a separate implementation, but would like to make use of the same plist/html stuff)

tanis2000 replied:
This does, indeed, make sense. I might as well start working on the S3 deployment as soon as I'm done rewriting this code to be less a hack and more a real implementation of deployment strategies :)

rayh commented on README.md:
```diff
@@ -79,10 +79,15 @@ This will produce something like: MyProject-Debug-1.0.ipa and MyProject-Debug-1.
-	end
+The function increment_build takes care of incrementing the build number of your product.
+If the version is just a number, it will increment that number.
+If the version is in dot notation (e.g. 1.2.3) it will increment the last value (leading to 1.2.4 in this example)
```
What if you want to increase the major/minor by some amount?  I think it would be better to add support for parsing and manipulating dotted version numbers.

Actually, I kind of think you should manage your marketing version manually (it doesnt really make sense to auto-increment it on every build).  And your internal bundle version is better off being inherited from your CI server.

I'd prefer to leave this sort of thing to the user to define in their own way rather than adding tools to deal with this specific scenario.  Could we back this change out?

tanis2000 replied:
It really depends on the project, at least in my own experience. There are situations where I'm not using a CI server and we're just pushing a new test version to the customer manually. In that scenario, we do increment both the bundle version and the last number (build) of the marketing version so that we know which build the customer is talking about when reporting bugs before it goes into production.

I agree that an implementation with full control over how the version is being incremented would be a way better option. To be honest I haven't looked around much for such a gem, maybe there already is something that does exactly that. I'll search the net and see if I can find something. It doesn't make sense to reinvent the wheel ;)

tanis2000 replied:
I actually found a gem that does exactly that. I'm rolling back those changes and switch to using that gem instead. 
Just in case you ever need it: https://github.com/dazuma/versionomy

rayh replied:
That's cool, maybe jot a point in the README about that gem as a suggestion
for how to manage versions

On 28 February 2013 02:58, Valerio Santinelli notifications@github.comwrote:

> In README.md:
> 
> > @@ -79,10 +79,15 @@ This will produce something like: MyProject-Debug-1.0.ipa and MyProject-Debug-1.
> > 
> >  ### Incrementing the build number
> > -   config.info_plist do |info|
> > -     info.version = info.version.to_i + 1
> > -     info.save
> > -   end
> >   +The function increment_build takes care of incrementing the build number of your product.
> >   +If the version is just a number, it will increment that number.
> >   +If the version is in dot notation (e.g. 1.2.3) it will increment the last value (leading to 1.2.4 in this example)
> 
> I actually found a gem that does exactly that. I'm rolling back those
> changes and switch to using that gem instead.
> Just in case you ever need it: https://github.com/dazuma/versionomy
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/rayh/xcoder/pull/56/files#r3172765
> .

## 

Ray Hilton

## (maker of mobile apps)

a: 13/243 Collins Street, Melbourne VIC 3000
m: 0430 484 708
w: http://rayh.com.au

rayh commented on lib/xcode/builder/base_builder.rb:
```diff
@@ -95,6 +95,23 @@ def testflight(api_token, team_token)
+      #
+      # @param protocol the protocol to use to upload files (ftp or ssh)
+      #
+      def web(protocol)
```
This is going to get unruly quite quickly... It would be better to keep the verbs at this level generic rather than service specific (i know testflight() contradict this, and this should change too).  Perhaps we have a method like:

``` ruby
def deploy method, options = {}
```

tanis2000 replied:
Ok, that's fine by me. I just followed the testflight example and put the method in there, but the way you suggest is cleaner.

It might take a while for me to figure out how to do this correctly as I'm new to Ruby development and I'm still learning the language.

rayh replied:
If you create a new deploy() method, I'll refactor the testflight stuff to
fit into that.  I just dont want to create new methods that we will only
take away again soon

On 27 February 2013 18:45, Valerio Santinelli notifications@github.comwrote:

> In lib/xcode/builder/base_builder.rb:
> 
> > @@ -95,6 +95,23 @@ def testflight(api_token, team_token)
> >          testflight.upload(ipa_path, dsym_zip_path)
> >        end
> > -      #
> > -      # Upload to web server
> > -      #
> > -      # The web object is yielded so further configuration can be performed before uploading
> > -      #
> > -      # @param protocol the protocol to use to upload files (ftp or ssh)
> > -      #
> > -      def web(protocol)
> 
> Ok, that's fine by me. I just followed the testflight example and put the
> method in there, but the way you suggest is cleaner.
> 
> It might take a while for me to figure out how to do this correctly as I'm
> new to Ruby development and I'm still learning the language.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/rayh/xcoder/pull/56/files#r3167139
> .

## 

Ray Hilton

## (maker of mobile apps)

a: 13/243 Collins Street, Melbourne VIC 3000
m: 0430 484 708
w: http://rayh.com.au

rayh commented:
Hi,

Thanks for this, I've added some comments, and I'd like to hear your thoughts.  But basically:
- Break the deployment class down into one-per-protocol
- Move HTML/Plist generation somewhere else (some utility classes, erb-backed or similar)
- Tests around what can be tested
- Documentation comments around new areas with examples, if possible

Also, I think the versioning stuff shouldn't be in this PR because:
a) it's quite specific to your projects (I, for example, don't do things that way).  Would be better as a separate project for parsing/generating marketing versions
b) marketing versions should probably not be auto incremented

Having said that, I should probably add some more explanation on suggested versioning strategies (i.e. the bundle version is internal, not for public consumption, so keep it traceable back to your CI, etc).  I only put the version + 1 in there as an example, but you should probably set it to ENV['BUILD_NUMBER'] if running under Jenkins.

tanis2000 commented:
Hi Ray,

thanks for the comments. I agree with you on all of them. I hacked those changes out of my inexperience with Ruby development and the need to have something that does the job quickly. Of course they can be rewritten in a better way.

I added some comments to your own and I suppose it should not be a problem to go over most of them except for the erb-backed stuff as I have no clue about how that works and I'll have to figure that out.

tanis2000 committed:
Rolled back versioning verbs

tanis2000 committed:
Mentioned Versionomy as a gem package to manage version numbers in the README

tanis2000 committed:
Trying some stuff for the deploy stage

tanis2000 committed:
First try at a generic deploy method with FTP protocol implementation

tanis2000 committed:
Better deploy method
SSH implementation of the deploy method
TODO: Still missing ERB generation of manifest.plist and index.html

tanis2000 committed:
First test with ERB

tanis2000 committed:
ERB generation of the manifest

tanis2000 committed:
Moved manifest template to ERB for SSH too

tanis2000 commented:
Hello Ray,

I've managed to implement a deploy method that can be used to run different deploy destinations. I've split up the web one to SSH and FTP but I haven't touched TestFlight as I'm not using it and I can't test it.

I've moved the manifest.plist and index.html generation to an ERB based one, too.

Now there still is a method called "prepare" that is exactly the same both in ftp.rb and ssh.rb and that should be refactored so that it's independent from the protocol, but this is more of a design decision and I'd like to hear from you where it should go.

Cheers,

Valerio

rayh commented:
Ok, well lets get this one merged and then look at refactoring internals... I'll update testflight as well.

Can you rebase and I'll take a look

tanis2000 commented:
Hi Ray, 

I have no idea how to rebase my changes. I kept working on my master branch and I have no clue at what I am supposed to do :)

rayh commented:
I'm going to merge the changes in, but we don't have any tests around the new stuff.   If you could then add some tests in the deploy_spec.rb and submit a new PR, that'll be lovely!

tanis2000 commented:
Ok, I'll get to the tests asap. Thanks!
